### PR TITLE
Thread related code commit

### PR DESCRIPTION
**Pull request by dfordivam**
This is a big commit which implements three major parts of torrent application 
1. Server thread
2. Control thread - For each torrent
3. Peer thread - For each peer communication

Currently these threads are just forked based on the tracker response. They dont do anything inside them. on Ctrl-C all these threads exit gracefully and torrent client stops.

### CODE REVIEW

File: src/FuncTorrent/Peer.hs [hunk 1]
```diff
@@ -1,59 +1,159 @@
 {-# LANGUAGE OverloadedStrings #-}
```
Comment (dfordivam) [inline src/FuncTorrent/Peer.hs:1]: These are changes from vu3rdd. But I guess its ok to merge them to master

File: src/Main.hs [hunk 2]
```diff
@@ -79,5 +85,27 @@ startTorrent log (m:_) = do
 
 startTorrent _ [] = return ()
 
+startTorrentConc :: (String -> IO ()) -> [Metainfo] -> IO ()
```
Comment (igniting) [inline src/Main.hs:88]: We can add a newtype for log.

File: src/Main.hs [hunk 3]
```diff
@@ -79,5 +85,27 @@ startTorrent log (m:_) = do
...
+  (st,_) <- initServerThread [(m,ct)]
+
+  -- Wait For user-interrupt
+  _ <- takeMVar interrupt
+
+  -- Exit gracefully
+  putMVar (st ^. serverTAction) FuncTorrent.ServerThread.Stop 
+  writeIORef (ct ^. controlTAction) FuncTorrent.ControlThread.Stop 
+  yield
+  threadDelay $ 4*1000*1000
```
Comment (randhir051) [inline src/Main.hs:108]: Replace with something better

File: src/Main.hs [hunk 4]
```diff
@@ -79,5 +85,27 @@ startTorrent log (m:_) = do
 
 startTorrent _ [] = return ()
 
+startTorrentConc :: (String -> IO ()) -> [Metainfo] -> IO ()
+startTorrentConc log (m:ms) = do
```
Comment (randhir051) [inline src/Main.hs:89]: Non exhaustive pattern match

File: test/ControlThreadTests.hs [hunk 5]
```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (randhir051) [inline test/ControlThreadTests.hs:1]: Add some module level documentation. As in one control thread per torrent or something like that

File: src/FuncTorrent/ControlThread.hs [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+{-# LANGUAGE TemplateHaskell #-}
+
+module FuncTorrent.ControlThread where
+
+import Control.Concurrent
+import GHC.Conc
+import Control.Lens
+import Data.IORef
+import Control.Monad hiding (
+    forM , forM_ , mapM , mapM_ , msum , sequence , sequence_ )
```
Comment (randhir051) [inline src/FuncTorrent/ControlThread.hs:11]: Space after , might be a lint error

Comment (randhir051) [inline src/FuncTorrent/ControlThread.hs:11]: Sort imports

Comment (igniting) [inline src/FuncTorrent/ControlThread.hs:11]: We can use stylish-haskell to sort imports

File: src/FuncTorrent/ControlThread.hs [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+-- 3. Stopping download/seed.
+--
+controlThreadMain :: ControlThread -> IO ()
+controlThreadMain ct =
+    doExit =<< (mainLoop <=< doInitialization) ct
+
+doInitialization :: ControlThread -> IO ControlThread
+doInitialization ct =
+  getTrackerResponse ct >>= \x ->
+  let peerInit = take 4 $ x^.peerList
```
Comment (randhir051) [inline src/FuncTorrent/ControlThread.hs:69]: Replace `x^.peerList` with `_peerList x`

File: src/FuncTorrent/ControlThread.hs [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  filterBadPeers ct >>=
+
+  pieceManagement >>=
+
+  -- Loop Here and check if we need to quit/exit
+  -- Add delay here before polling PeerThreads again
+  checkAction
+
+ where
+   checkAction ct1 = do
```
Comment (randhir051) [inline src/FuncTorrent/ControlThread.hs:85]: Add type signature for check action

File: src/FuncTorrent/ControlThread.hs [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+     putStrLn "Check action"
+     threadDelay $ 4*1000*1000
```
Comment (randhir051) [inline src/FuncTorrent/ControlThread.hs:87]: RED FLAG

File: src/FuncTorrent/ControlThread.hs [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+     action <- readIORef $ view controlTAction ct1
+     case action of
+       FuncTorrent.ControlThread.Stop -> return ct1
+       _ -> mainLoop ct1
+
+doExit :: ControlThread -> IO ()
```
Comment (randhir051) [inline src/FuncTorrent/ControlThread.hs:93]: Add doc.

**Comment by jaseemabid** [comment]
@dfordivam Can you check why the travis build is failing? Merge if it works.

**Comment by jaseemabid** [comment]
@dfordivam 

test/PeerThreadMock.hs:44:33: Not in scope: ‘status’
test/PeerThreadMock.hs:45:34: Not in scope: ‘action’

Fix the tests.

File: functorrent.cabal [hunk 11]
```diff
@@ -81,3 +91,31 @@ test-suite functorrent-test
                      QuickCheck,
                      tasty-quickcheck,
                      test-framework-quickcheck2
+
+test-suite control-thread-tests
```
Comment (jaseemabid) [inline functorrent.cabal:95]: A new test suite is not required. Cant you add a few more tests to the existing quite?

File: src/FuncTorrent/ControlThread.hs [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+--    ,   _diskIO_Handle   :: Handle
+    ,   _controlTStatus     :: IORef ControlThreadStatus
+    ,   _controlTAction     :: IORef ControlThreadAction
+    }
+
+data ControlThreadStatus =
+        Stopped
+    |   Downloading
+    |   Seeding
+  deriving (Eq, Show)
```
Comment (jaseemabid) [inline src/FuncTorrent/ControlThread.hs:35]: Any particular reason for `deriving (Eq, Show)` ? Same question about `ControlThreadAction`

File: src/FuncTorrent/ControlThread.hs [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+data ControlThreadAction =
+        Download
+    |   Pause
+    |   Seed
+    |   Stop
+  deriving (Eq, Show)
+
+makeLenses ''ControlThread
```
Comment (jaseemabid) [inline src/FuncTorrent/ControlThread.hs:44]: This as per we agreed in the last meeting, we will let this PR land, but will remove all lenses business soon after that.

File: src/FuncTorrent/ControlThread.hs [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+  case decode response of
+    Right trackerInfo ->
+        case mkTrackerResponse trackerInfo of
+          Right trackerResp ->
+            let ct1 = trackerResponses %~ (trackerResp : ) $ ct
+                ct2 = peerList %~ ((peers trackerResp) ++) $ ct1
+            in return ct2
+          Left _ -> putStrLn "mkTracker error" >> return ct
+    Left _ -> putStrLn "tracker resp decode error" >> return ct
+
```
Comment (jaseemabid) [inline src/FuncTorrent/ControlThread.hs:138]: Both left cases seems wrong. Silent errors.

File: functorrent.cabal [hunk 15]
```diff
@@ -23,9 +23,14 @@ library
...
                        FuncTorrent.Tracker
+                       FuncTorrent.ControlThread
+                       FuncTorrent.PeerThread
+                       FuncTorrent.PeerThreadMain
+                       FuncTorrent.PeerThreadData
+                       FuncTorrent.ServerThread
 
   other-modules:       FuncTorrent.Utils
-  other-extensions:    OverloadedStrings
+  other-extensions:    OverloadedStrings, CPP
```
Comment (vu3rdd) [inline functorrent.cabal:33]: Is CPP needed for the #ifdefine and friends?

**Pull request merged**